sigs/jvm/rfcs/20190904-java-tensors.md

```diff
@@ -0,0 +1,301 @@
+read a `UInt8` value from memory, you actually read a `Byte`. 
+
+Since the parameter of the `NdArray` must be the Java type of the data stored in memory, since `Tensor` needs
+to implement the `NdArray` interface to allow direct I/O operations, and since we carry the TF data type as a
+parameter to `Tensor` for operand type safety, we would end up needing two generic types in the `Tensor`
+signature, e.g. `Tensor<UInt8, Byte>` or `Tensor<Integer, Integer>`. 
+
+This can start to be annoying pretty fast. The current proposal main purpose is to avoid the hassle of carrying
+to much information in generic parameters by taking a completely different approach.
+
+### Tensor Types
+
+#### Tensor Classes
+
+A solution is to identify tensor data types not by a type parameter but by a concrete class. There would be
+one class per supported data type, which extends from the base class `Tensor<T>`, where `T` is the Java type
+used for data access (therefore is used to implement `NdArray<T>`).
+
+In addition, having our own type concrete classes allows us to group those types into families that can be
+used to add more contraints on which type of tensor is allowed for a given TF operation (e.g. only `Numeric`
+data types can be used in a `tf.math.add` operation).
+
+Let see what it looks like in a quick example:
+
+```java
+public abstract class Tensor<T> implements NdArray<T> { ... }
+
+public interface Numeric {}
+
+public final class TInt32 extends Tensor<Integer> implements Numeric { ... }
+
+public final class TUInt8 extends Tensor<Byte> implements Numeric { ... }
+
+public final class TString extends Tensor<String> { ... }
+```
+
+To continue to support compile-time type safety in the operands of a TF operations, we carry that concrete type name
+as a parameter to the operations (instead of the actual idiomatic aliases).
+
+```java
+Constant<TInt32> c = tf.constant(2); // more on this constant creation in the next section...
+
+Add<TInt32> addOp = tf.math.add(c, c);
+
+TInt32 addResult = addOp.tensor(); // only possible in eager mode
+
+Integer sum = addResult.get(); // from NdArray<Integer> interface
+```
+
+The last lines are interesting, as in eager mode, we can easily convert the result of an operation to a tensor
+of the appropriate type and then read its data. We can also achieve something similar when fetching results
+of a graph session.
+
+#### Data Types
+
+Now even with those new concrete classes, we need to be able to carry information about a given data type without
+accessing a tensor instance. Actually, this information is carried by the [`DataType` enum class](https://github.com/karllessard/tensorflow/blob/master/tensorflow/java/src/main/java/org/tensorflow/DataType.java),
+which converts back and forth a type alias (such as `Integer`) to a TF data type (such as `INT32`).
+
+Again, this conversion can be sometimes painful and could be avoid by changing the `DataType` enum class to a 
```

Comment:
Ok, maybe I mixed two different concept here.

The main idea was not to get rid of the enum but to get rid of the conversion between a Java type and an enum, using [this method](https://github.com/tensorflow/tensorflow/blob/a2048b0a19157054a6bca25610c8b7aa14bef2d5/tensorflow/java/src/main/java/org/tensorflow/DataType.java#L96), like [here](https://github.com/tensorflow/tensorflow/blob/a2048b0a19157054a6bca25610c8b7aa14bef2d5/tensorflow/java/src/main/java/org/tensorflow/op/core/Zeros.java#L50).

Now why using a static class instead of an enum? Well, since we already have a specific class per supported tensor type (in this RFC I mean), for me it sounds just more intuitive to store all information related to that type inside this class and the user does not have to deduce the corresponding value in an enum. (e.g. the type info of `TInt32`is `TInt32.DTYPE`, instead of the type of of `TInt32`is `DataType.INT32`).

Reply:
Ok. That approach precludes switching over the types or other things which need to know about all the Tensor types in the C API. Which might be fine, as you point out later it's much more extensible, but I'm not sure if it's worth exposing types in Java that aren't understood by the C API underneath.

Reply:
> but I'm not sure if it's worth exposing types in Java that aren't understood by the C API underneath.

That is not the plan neither, all tensor class such a `TInt32`would match a type supported by the C API. Or maybe you mean the "custom" type I was proposing later? Yes, this one need more rethinking for sure.

Reply:
Yeah, I mean the custom types proposed at the end of the document.

Reply:
However there are things like fp16 and bf16 which aren't easily representable in Java at the moment, but are probably useful.